cpp/src/slice2java/Gen.h

```diff
@@ -8,89 +8,129 @@
 
 namespace Slice
 {
+    //
+    // Generate code to marshal or unmarshal a type.
+    //
+    enum OptionalMode
+    {
+        OptionalNone,
+        OptionalInParam,
+        OptionalOutParam,
+        OptionalReturnParam,
+        OptionalMember
+    };
+
     class JavaVisitor : public JavaGenerator, public ParserVisitor
     {
     public:
         ~JavaVisitor() override;
 
-    protected:
-        JavaVisitor(const std::string&);
```

Comment:
This `protected` section was moved to the bottom of the class.
There is no reason for all the helper functions to be protected instead of plain 'ol public here.